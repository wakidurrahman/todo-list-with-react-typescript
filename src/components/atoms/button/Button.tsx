/**
 * A reusable Button component that follows Bootstrap styling conventions.
 * Supports different variants, sizes, and states while maintaining accessibility.
 */

type ButtonProps = {
  /** Function to be called when the button is clicked */
  onClick?: () => void;
  /** Variant of the button, defaults to 'primary' */
  variant?: 'primary' | 'secondary' | 'danger' | 'success';
  /** Size of the button, defaults to 'md' */
  size?: 'sm' | 'md' | 'lg';
  /** Content of the button */
  children?: React.ReactNode;
  /** Type of the button, defaults to 'button' */
  type?: 'button' | 'submit';
  /** Whether the button is disabled */
  disabled?: boolean;
  /** Additional CSS classes to apply to the button */
  className?: string;
};

function Button({
  onClick,
  variant = 'primary',
  size = 'md',
  children,
  type = 'button',
  disabled = false,
  className = '',
  ...rest
}: ButtonProps) {
  // Combine base classes with additional classes
  const baseClass = `btn btn-${variant} btn-${size} ${className}`.trim();

  return (
    <button
      type={type}
      className={baseClass}
      onClick={onClick}
      disabled={disabled}
      {...rest}
    >
      {children}
    </button>
  );
}

export default Button;
