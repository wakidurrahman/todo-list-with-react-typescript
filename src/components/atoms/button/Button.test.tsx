import { render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';

import Button from './Button';

describe('Button Component', () => {
  // Snapshot Tests
  describe('Snapshots', () => {
    it('renders primary button correctly', () => {
      const { container } = render(
        <Button variant="primary" onClick={() => {}}>
          Primary Button
        </Button>
      );
      expect(container).toMatchSnapshot();
    });

    it('renders secondary button correctly', () => {
      const { container } = render(
        <Button variant="secondary" onClick={() => {}}>
          Secondary Button
        </Button>
      );
      expect(container).toMatchSnapshot();
    });

    it('renders disabled button correctly', () => {
      const { container } = render(
        <Button variant="primary" disabled onClick={() => {}}>
          Disabled Button
        </Button>
      );
      expect(container).toMatchSnapshot();
    });
  });

  // Unit Tests
  describe('Functionality', () => {
    const user = userEvent.setup();

    it('calls onClick handler when clicked', async () => {
      const handleClick = jest.fn();
      render(
        <Button variant="primary" onClick={handleClick}>
          Click Me
        </Button>
      );

      const button = screen.getByRole('button', { name: /click me/i });
      await user.click(button);

      expect(handleClick).toHaveBeenCalledTimes(1);
    });

    it('does not call onClick when disabled', async () => {
      const handleClick = jest.fn();
      render(
        <Button variant="primary" onClick={handleClick} disabled>
          Disabled Button
        </Button>
      );

      const button = screen.getByRole('button', { name: /disabled button/i });
      await user.click(button);

      expect(handleClick).not.toHaveBeenCalled();
    });

    it('applies correct Bootstrap classes based on variant', () => {
      const { rerender } = render(
        <Button variant="primary" onClick={() => {}}>
          Primary Button
        </Button>
      );

      let button = screen.getByRole('button');
      expect(button).toHaveClass('btn', 'btn-primary');

      rerender(
        <Button variant="secondary" onClick={() => {}}>
          Secondary Button
        </Button>
      );

      button = screen.getByRole('button');
      expect(button).toHaveClass('btn', 'btn-secondary');
    });

    it('applies additional className when provided', () => {
      render(
        <Button variant="primary" onClick={() => {}} className="custom-class">
          Custom Button
        </Button>
      );

      const button = screen.getByRole('button');
      expect(button).toHaveClass('custom-class');
    });
  });
});
