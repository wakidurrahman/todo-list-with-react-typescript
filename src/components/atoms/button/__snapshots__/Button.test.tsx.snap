// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button Component Snapshots renders disabled button correctly 1`] = `
<div>
  <button
    class="btn btn-primary btn-md"
    disabled=""
    type="button"
  >
    Disabled Button
  </button>
</div>
`;

exports[`Button Component Snapshots renders primary button correctly 1`] = `
<div>
  <button
    class="btn btn-primary btn-md"
    type="button"
  >
    Primary Button
  </button>
</div>
`;

exports[`Button Component Snapshots renders secondary button correctly 1`] = `
<div>
  <button
    class="btn btn-secondary btn-md"
    type="button"
  >
    Secondary Button
  </button>
</div>
`;
